peglib.h

```diff
@@ -1295,25 +1341,104 @@ class LiteralString : public Ope
+        // \s matches whitespace (short for [ \f\n\r\t\v\u00A0\u2028\u2029]).
+        ranges_.emplace_back(std::make_pair(' ', ' '));
+        ranges_.emplace_back(std::make_pair('\f', '\f'));
+        ranges_.emplace_back(std::make_pair('\n', '\n'));
+        ranges_.emplace_back(std::make_pair('\r', '\r'));
+        ranges_.emplace_back(std::make_pair('\t', '\t'));
+        ranges_.emplace_back(std::make_pair('\v', '\v'));
+        ranges_.emplace_back(std::make_pair(0xA0, 0xA0));
+        ranges_.emplace_back(std::make_pair(0x2028, 0x2028));
+        ranges_.emplace_back(std::make_pair(0x2029, 0x2029));
+        specialChars[pos] = std::make_pair("\\s", 9);
+    }
+
+    void insertWordMatches(size_t pos) {
+        ranges_.emplace_back(std::make_pair('_', '_'));
+        ranges_.emplace_back(std::make_pair('a', 'z'));
+        ranges_.emplace_back(std::make_pair('A', 'Z'));
+        ranges_.emplace_back(std::make_pair('0', '9'));
+        specialChars[pos] = std::make_pair("\\w", 4);
+    }
+
+    void fillfromString(const std::string &str) {
+        auto chars = decode(str.c_str(), str.length());
+        for (size_t i = 0u; i < chars.size(); ++i) {
+            if (i + 2 < chars.size() && chars[i + 1] == '-' && chars[i] != '\\') {
                 auto cp1 = chars[i];
                 auto cp2 = chars[i + 2];
                 ranges_.emplace_back(std::make_pair(cp1, cp2));
-                i += 3;
+                i += 2;
+            } else if (i+1 < chars.size() && chars[i] == '\\') {
+                // subset of regex shorthands
+                // http://www.javascriptkit.com/javatutors/redev2.shtml
+                ++i;
+                switch (chars[i]) {
+                case '-': case '^': case '[': case ']':
+                    --i;
+                    break;
+                case 's':
+                    insertWhiteSpace(i-1); break;
+                case 'w':
+                    insertWordMatches(i-1); break;
+                case 'd':
+                    ranges_.emplace_back(std::make_pair('0', '9'));
+                    specialChars[i-1] = std::make_pair("\\d", 1);
+                    break;
+                case '\\': ranges_.emplace_back(std::make_pair('\\', '\\')); break;
+                default: {
+                    std::string message = "Unsupported escape char '\\";
+                    message += static_cast<const char>(chars[i]);
+                    message += "'";
+                    throw parse_error(message.c_str());
+                }
+                }
             } else {
                 auto cp = chars[i];
                 ranges_.emplace_back(std::make_pair(cp, cp));
-                i += 1;
             }
         }
     }
 
-    CharacterClass(const std::vector<std::pair<char32_t, char32_t>>& ranges) : ranges_(ranges) {}
+public:
+    CharacterClass(const std::string& str)
+    {
+        // first char in char class might negate char class
+        negated = str.front() == '^';
```

Comment:
'negated' character class support seems to be common in many PEG libraries. But this should be recognized in the PEG grammar.